pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -274,3 +285,26 @@ def publish(self, topic, data, **attrs):
                 batch = self._batch(topic, create=True)
 
```

Comment:
I believe the stopping lock still leaves a race in this code. Ex:

Thread 1: client.publish() starts running, checks _is_stopped and keeps going because it's False
Thread 2: client.stop() runs, sets _is_stopped to True, and calls batch.commit() on all batches
Thread 1: continues client.publish(), adds a batch, calls batch.publish(), and returns a future to the published message

Now, there's a dangling message publish that stop() has not committed.

One option is to take the _is_stopped lock for the entire duration of publish() and stop(). We'd be extending critical sections for a long time though..

Another is to 1) remove the _is_stopped lock 2) take the batching lock in stop() for the whole method 3) remove the batching lock from client._batch() 4) insert the batching lock between lines 280 and 285 in client.publish() (before batch = self._batch(topic)" and after "batch = self._batch(topic, create=True)". This option is simplest and fixes the issue. We extend the critical section only a little bit. In any case, I believe extending the critical section like this will be necessary for future ordering keys work.

What do you think?

Reply:
> Another is to 1) remove the _is_stopped lock...

Sounds reasonable, though I don't think we need to take lock for entire `stop()`, as we are preventing further publishing with `_is_stopped`. If publish started, `stop()` will wait for publish completion. And if `stop()` started earlier, publish can be interrupted right after setting `_is_stopped` to `True` - there is no need to wait for commits completion.

Pushed changes.

Reply:
I think we still need a batch lock in stop() because commit() will race with batch creation & publish otherwise. Consider this ordering:

T1: client.publish() runs, sees is_stopped is False, and continues.
T2: sets is_stopped to True, runs commit() on all existing batches
T1: still in publish(), adds message to new/old batch and publishes. This causes a message addition to a committed batch (this results in batch.publish() returning a None future, causing a new batch to be created and not published immediately- a dangling batch), or starts a new batch that is then published (this new batch has to wait for the commit timer to commit - a dangling batch).

Reply:
Yes, I've added lock on half of `stop()`, while `publish()` is completely locked. So if `publish()` started first, it'll deal with batches, and only after that `stop()` will start to do it's job. But if `stop()` started first, it'll change `is_stopped` to `True` with lock, so further `publish()` will no go.

Reply:
Ah! Sorry, I misread this change. The locking behavior is correct as is. I do think it'll be simpler if we just lock around all of client.stop(). It reduces the number of orderings one must consider while reading. It will have little performance impact since client.stop() should be seldom called. What do you think?

Reply:
Makes some sense. After all, we're stopping the publishing. Pushed

Reply:
Thanks!